### PR TITLE
Unlimited Split-Screen

### PR DESCRIPTION
**Pull request by Fantasmos**
Allows for unlimited splitscreen functionality. 

Fixes have been done for scaling GUI elements (Message, Map, Player-icons, lapcount).

Spinners should be fixed, right now it just loops the first 4 colours which is probably sufficient.

FOV is set to be the same as if you were 4 players for when there's more than 4 people.

Added a new option in the UI section to allow for horizontal and vertical split-screen. 

Another issue worth investigating is if floating point loss causes higher playercounts to make black lines at the right edge of the screen (that are like, 3 pixels wide)

### CODE REVIEW

**Comment by hiker** [comment]
5-way-splitscreen
A picture from @Fantasmos

**Comment by Fantasmos** [comment]
By the way this picture is out of date, the FOV is more consistent and appropriate now. I'll try and make a full album tonight

File: src/config/stk_config.cpp [hunk 1]
```diff
@@ -268,10 +268,12 @@ void STKConfig::getAllData(const XMLNode * root)
 
     if (const XMLNode *camera = root->getNode("camera"))
     {
-        camera->get("fov-1", &m_camera_fov[0]);
-        camera->get("fov-2", &m_camera_fov[1]);
-        camera->get("fov-3", &m_camera_fov[2]);
-        camera->get("fov-4", &m_camera_fov[3]);
+        for (int i = 0; i < 4; i++) {
```
Comment (auriamg) [inline src/config/stk_config.cpp]: Our coding convention is for braces to go on their own line

File: src/graphics/irr_driver.cpp [hunk 2]
```diff
@@ -251,7 +251,25 @@ void IrrDriver::updateConfigIfRelevant()
     }
 #endif   // !SERVER_ONLY
 }   // updateConfigIfRelevant
-
+core::recti IrrDriver::GetSplitscreenWindow(int WindowNum) {
+    const int playernum = race_manager->getNumLocalPlayers();
+    const float Sqrt = sqrt(playernum);
```
Comment (auriamg) [inline src/graphics/irr_driver.cpp]: our convention for variables is that they should be lower_case

File: src/graphics/irr_driver.cpp [hunk 3]
```diff
@@ -251,7 +251,25 @@ void IrrDriver::updateConfigIfRelevant()
...
+	
+    const int rows = UserConfigParams::Split_Screen_Horizontally ? ceil(Sqrt) : round(Sqrt);
+    const int cols = UserConfigParams::Split_Screen_Horizontally ? round(Sqrt) : ceil(Sqrt);
+
+    const int width_of_space = floor(irr_driver->getActualScreenSize().Width / cols);
+    const int height_of_space = floor(irr_driver->getActualScreenSize().Height / rows);
+
+    const int X_Grid_Position = WindowNum % cols;
```
Comment (auriamg) [inline src/graphics/irr_driver.cpp]: our convention for variables is that they should be lower_case

File: src/guiengine/skin.cpp [hunk 4]
```diff
@@ -1271,9 +1271,10 @@ void Skin::drawSpinnerBody(const core::recti &rect, Widget* widget,
             params=&SkinConfig::m_render_params["spinner4::neutral"];
         else
         {
-            Log::fatal("Skin::drawSpinnerBody", "Unknown playerID (more than 4 players?)");
+            //TODO Add individual spinner skins
```
Comment (auriamg) [inline src/guiengine/skin.cpp]: This TODO will have to be resolved before this is merged

File: src/states_screens/options_screen_ui.cpp [hunk 5]
```diff
@@ -128,7 +128,9 @@ void OptionsScreenUI::init()
     assert( skinSelector != NULL );
 
     // ---- video modes
-
+	CheckBoxWidget* SplitScreen_Method = getWidget<CheckBoxWidget>("Split_Screen_Horizontally");
```
Comment (auriamg) [inline src/states_screens/options_screen_ui.cpp]: Our convention is to indent with spaces, no tabs

File: src/guiengine/skin.cpp (hunk 4 above)
Comment (Fantasmos) [reply src/guiengine/skin.cpp]: I don't think it should come in the way of a merge, but definitely should be left as an issue, or something on an agenda to do for the next release milestone

File: src/states_screens/options_screen_ui.cpp [hunk 6]
```diff
@@ -258,6 +260,12 @@ void OptionsScreenUI::eventCallback(Widget* widget, const std::string& name, con
         UserConfigParams::m_skin_file = core::stringc(selectedSkin.c_str()).c_str() + std::string(".stkskin");
         GUIEngine::reloadSkin();
     }
+	else if (name == "Split_Screen_Horizontally")
```
Comment (auriamg) [inline src/states_screens/options_screen_ui.cpp]: Please use spaces to indent, not tabs

File: src/states_screens/options_screen_ui.cpp [hunk 7]
```diff
@@ -258,6 +260,12 @@ void OptionsScreenUI::eventCallback(Widget* widget, const std::string& name, con
...
+	{
+		CheckBoxWidget* Split_Screen_Horizontally = getWidget<CheckBoxWidget>("Split_Screen_Horizontally");
```
Comment (auriamg) [inline src/states_screens/options_screen_ui.cpp]: our convention for variable names is lower-case

File: src/states_screens/race_gui.cpp [hunk 8]
```diff
@@ -888,14 +887,18 @@ void RaceGUI::drawLap(const AbstractKart* kart,
...
     // If the time display in the top right is in this viewport,
     // move the lap/rank display down a little bit so that it is
     // displayed under the time.
-    if (viewport.UpperLeftCorner.Y==0 &&
-        viewport.LowerRightCorner.X==(int)(irr_driver->getActualScreenSize().Width) &&
-        race_manager->getNumPlayers()!=3)
-        pos.UpperLeftCorner.Y   += m_font_height;
+    if (viewport.UpperLeftCorner.Y == 0 &&
+        viewport.LowerRightCorner.X == (int)(irr_driver->getActualScreenSize().Width) &&
+        !race_manager->getIfEmptyScreenSpaceExists()) {
```
Comment (auriamg) [inline src/states_screens/race_gui.cpp]: The brace goes on a newline

File: src/states_screens/race_gui.cpp [hunk 9]
```diff
@@ -888,14 +887,18 @@ void RaceGUI::drawLap(const AbstractKart* kart,
...
+        pos.UpperLeftCorner.Y += m_font_height;
+    }
+    else {
```
Comment (auriamg) [inline src/states_screens/race_gui.cpp]: The brace goes on a newline

File: data/gui/options_ui.stkgui [hunk 10]
```diff
@@ -72,8 +72,12 @@
                 </div>
 
                 <spacer height="4" width="10" />
-
-                <!-- Nothing here yet -->
+                
+                <div layout="horizontal-row" proportion="1" height="fit">
+                    <checkbox id="Split_Screen_Horizontally"/>
+                    <spacer width="20" height="100%" />
+                    <label height="100%" I18N="In the ui settings" text="Split Screen horizontally (At next race)"/>
```
Comment (auriamg) [inline data/gui/options_ui.stkgui]: To be honest I have no idea how to improve that, but as a regular player, if I just saw that option, I don't think I would understand what it does :/ Not sure if anyone has ideas on how to make this clearer

**Comment by Fantasmos** [comment]
Added a fix to this PR for the potential 'dead' pixels on edges due to rounding. Will need to test it though just to be sure (however I never tested the issue in the first place)

An example of the issue was having 7 players at 1280x720 causes their width to be 426 pixels across 3 columns, which lead to a total width only of 1278. However now we will use a ceiling instead of floor, and restrict results that are larger than the total screen width.

**Comment by auriamg** [comment]
@Fantasmos Is this finished? There's been quite a few recent commits

**Comment by Fantasmos** [comment]
Needs a test, but in theory it's finished

**Comment by auriamg** [comment]
@Fantasmos well I guess it's no longer finished since you added a new WIP commit to the PR just after telling me it was finished ;) I thought maybe I could have merged the rest before going there

**Comment by Fantasmos** [comment]
@auriamg Thanks for the understanding, these last few PR's have been a lot more rocky than anticipated. 

Good news is however, this PR (Apart from design decisions and styling) is finished. I had to rework a bit of the code that handles how 'selected' objects are hovered over in order to get multiplayer selection working programatically but the difference should not be apparent. 

I'll hopefully clean this up, and discuss some of the small issues with you all soon.

Small issues are pretty much: 
-On the character selection screen, Player1's selector is still the regular selector instead of being based on color. 
-In some skins (forest demonstrates this best) this clash causes Player1 to look too similar to other players. 

These issues I find to be negligible compared to having 8 player split-screen functionality, and the potential hazard of modifying pre-existing code for this special use case.

**Comment by Fantasmos** [comment]
https://www.youtube.com/watch?v=KQT9qsLeLYM
This video should a lot that has been modified, or needs to be considered with splitscreen, hope it can make the process easier.

**Comment by qwertychouskie** [comment]
@Fantasmos I get "This video is unavailable" when clicking the link.

**Comment by Fantasmos** [comment]
Thanks, accidentally had it private not unlisted. Fixed it now

**Pull request merged**